varnamc.c

```diff
@@ -0,0 +1,379 @@
+static char args_doc[] = "";
+static struct argp_option options[] = { 
+    {"symbols", 's', "VALUE", 0, "Sets the symbols file"},
+    {"transliterate", 't', "TEXT", 0, "Transliterate the given text"},
+    {"details", 'd', "", OPTION_ARG_OPTIONAL, "Detailed transliteration output. Use with --transliterate"},
+    {"reverse-transliterate", 'r', "TEXT", 0, "Reverse transliterate the given text"},
+    {"learn", 'n', "TEXT", 0, "Learn the given text"},
+    {"learn-from", 'f', "FILE", 0, "Reads from the specified file"},
+    {"train", 'a', "PATTERN=WORD", 0, "Train the given text"},
+    {"import-learnings-from", 'i', "FILE", 0, "Import learned data from the specified file"},
+    {"version", 'v', "", OPTION_ARG_OPTIONAL, "Display version"},
+    {0} 
+};
+
+struct arguments {
+  char *symbols;
+  char *transliterate;
+  bool details;
+  char *reverse_transliterate;
+  char *learn;
+  char *learn_from;
+  char *train;
+  char *import_learnings_from;
+  bool version;
+};
+
+static error_t parse_opt(int key, char *arg, struct argp_state *state) {
+  struct arguments *arguments = state->input;
+  switch (key) {
+  case 's':
+    arguments->symbols = arg;
+    break;
+  case 't':
+    arguments->transliterate = arg;
+    break;
+  case 'd':
+    arguments->details = true;
+    break;
+  case 'r':
+    arguments->reverse_transliterate = arg;
+    break;
+  case 'n':
+    arguments->learn = arg;
+    break;
+  case 'f':
+    arguments->learn_from = arg;
+    break;
+  case 'a':
+    arguments->train = arg;
+    break;
+  case 'i':
+    arguments->import_learnings_from = arg;
+    break;
+  case 'v':
+    arguments->version = true;
+    break;
+  case ARGP_KEY_ARG:
+    return 0;
+  default:
+    ARGP_ERR_UNKNOWN;
+  }
+  return 0;
+}
+
+static struct argp argp = { options, parse_opt, args_doc, doc, 0, 0, 0 };
+
+/**
+ * -----
+ * Helper Functions
+ * -----
+ */
+
+/* Check if text is not a sentence */
+void ensure_single_word(char *text)
+{
+  int i = 0;
+  size_t s_length = strlen(text);
+  while(i < s_length)
+  {
+    if(text[i++] == ' ')
+    {
+      printf("varnamc : Expected a single word.");
+      exit(1);
+    }
+  }
+}
+
+
+/**
+ *  splits str on delim and dynamically allocates an array of pointers.
+ *
+ *  On error -1 is returned, check errno
+ *  On success size of array is returned, which may be 0 on an empty string
+ *  or 1 if no delim was found.  
+ *
+ *  You could rewrite this to return the char ** array instead and upon NULL
+ *  know it's an allocation problem but I did the triple array here.  Note that
+ *  upon the hitting two delim's in a row "foo,,bar" the array would be:
+ *  { "foo", NULL, "bar" } 
+ * 
+ *  You need to define the semantics of a trailing delim Like "foo," is that a
+ *  2 count array or an array of one?  I choose the two count with the second entry
+ *  set to NULL since it's valueless.
+ *  Modifies str so make a copy if this is a problem
+ */
+int split( char * str, char delim, char ***array )
```

Comment:
https://github.com/antirez/sds

could you try this library. check weather this helps

Reply:
The Tokenization part in that SDS library is exactly what we need. So, C libraries are used just by copy-paste ah ? Perhaps I should use git submodules for better dependency injection.